Add ONNX support for Pegasus

pramodith opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

This PR adds ONNX support for the Pegasus model. Linked to https://github.com/huggingface/transformers/issues/16308


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed.
@lewtun 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

pramodith committed:
Adding changes to transformers library to include Onnx Config for the Pegasus model.

pramodith committed:
Merge branch 'huggingface:main' into pramodith_pegasus_onnx_config

pramodith committed:
Correcting typo in features for the name of the PegasusOnnx class.

pramodith committed:
Correcting typo in features for the name of the PegasusOnnx class.

pramodith committed:
Merge remote-tracking branch 'origin/pramodith_pegasus_onnx_config' into pramodith_pegasus_onnx_config

pramodith committed:
Removing unused import in file.

pramodith committed:
Removing unused import in file.

pramodith committed:
Removing unused import in file.

pramodith committed:
Changes after running all the make commands to correct formatting and update the documentation.

pramodith committed:
Changes after running all the make commands to correct formatting and update the documentation.

pramodith committed:
Changes after running all the make commands to correct formatting and update the documentation.

pramodith committed:
Changes after running all the make commands to correct formatting and update the documentation.

pramodith committed:
Changes after running all the make commands to correct formatting and update the documentation.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18305). All of your documentation changes will be reflected on that endpoint.

ChainYo commented:
Hey @pramodith, this PR looks excellent! Did you try to convert a `Pegasus` model with this config?

I'm pinging @lewtun for reviewing.

pramodith commented:
@ChainYo  besides the pytests, I also tried executing the Onnx export and validation using this script. The difference between the pytorch tensors and the Onnx tensors was in the order of 1e-5. However, I did notice that the onnx export would fail when I set use_past=True, not sure if I need to do anything about that.

```from transformers import AutoConfig, AutoTokenizer, PegasusForCausalLM,PegasusForConditionalGeneration,PegasusModel
from transformers.models.pegasus import PegasusOnnxConfig
from transformers.onnx import export, validate_model_outputs
from pathlib import Path

def check_onnx_model(task):
    if task == "default":
        config = AutoConfig.from_pretrained("google/pegasus-large")
        model = PegasusModel.from_pretrained("google/pegasus-large")
        tokenizer = AutoTokenizer.from_pretrained("google/pegasus-large")
    elif task == "seq2seq-lm":
        config = AutoConfig.from_pretrained("google/pegasus-xsum")
        model = PegasusForConditionalGeneration.from_pretrained("google/pegasus-xsum", add_cross_attention=True)
        tokenizer = AutoTokenizer.from_pretrained("google/pegasus-xsum")
    else:
        config = AutoConfig.from_pretrained("google/pegasus-xsum")
        model = PegasusForCausalLM.from_pretrained("google/pegasus-xsum")
        tokenizer = AutoTokenizer.from_pretrained("google/pegasus-xsum")

    onnx_config = PegasusOnnxConfig(config, task=task, use_past=True)
    onnx_path = Path("model.onnx")
    onnx_inputs, onnx_outputs = export(tokenizer, model, onnx_config, onnx_config.default_onnx_opset, onnx_path)
    print(onnx_inputs)
    print(onnx_outputs)
    print(validate_model_outputs(onnx_config,tokenizer,model,onnx_path,onnx_outputs,onnx_config.atol_for_validation))

check_onnx_model("causal-lm")
```

pramodith commented:
> > @ChainYo, besides the pytests, I also tried executing the Onnx export and validation using this script. The difference between the PyTorch tensors and the Onnx tensors was in the order of 1e-5. However, I did notice that the onnx export would fail when I set use_past=True, not sure if I need to do anything about that.
> 
> I'm not sure you can use past when you convert your model to ONNX. You have to rewrite the logic when using that feature with ONNX.
> 
> **Edit**: I never converted a model which uses `past`, but it could work if your config inherits from `OnnxSeq2SeqConfigWithPast`, which is the case.

Gotcha! I'll wait for any comments by the reviewers in that case.

ChainYo commented:
Hi, I just checked the failed test in the CI/CD pipeline, and it doesn't seem to come from your code. Could you please try to fetch/rebase your branch with the main upstream branch? We will see if it's solved.

[CI/CD error](https://app.circleci.com/pipelines/github/huggingface/transformers/44492/workflows/4da69f9f-f332-4ba8-97e7-3446ec6a559e/jobs/517244?invite=true#step-111-3911)

pramodith committed:
Rebasing to align with upstream branch.

pramodith committed:
Rebasing to align with upstream branch.

pramodith committed:
Rebasing to align with upstream branch.

pramodith committed:
Merge branch 'main' into pramodith_pegasus_onnx_config

pramodith commented:
@ChainYo I've rebased and all the tests pass now!

ChainYo commented:
> @ChainYo I've rebased, and all the tests pass now!

Super let's just wait for a reviewer now!

lewtun commented on tests/onnx/test_onnx_v2.py:
```diff
@@ -230,6 +230,7 @@ def test_values_override(self):
     ("blenderbot", "facebook/blenderbot-400M-distill"),
     ("bigbird-pegasus", "google/bigbird-pegasus-large-arxiv"),
     ("longt5", "google/long-t5-local-base"),
+    ("pegasus", "google/pegasus-large"),
```
Do you happen to know if there exists a smaller checkpoint on the Hub that can serve for testing?

lewtun commented on src/transformers/models/pegasus/configuration_pegasus.py:
```diff
@@ -167,3 +171,74 @@ def num_attention_heads(self) -> int:
         return self.d_model
+
+
+class PegasusOnnxConfig(OnnxSeq2SeqConfigWithPast):
```
Since Pegasus is based on BART, I think we can copy-paste the config implemented in `configuration_bart.py` and add a `# Copied from transformers.models.bart.configuration_bart.BartOnnxConfig with Bart->Pegasus`

This should handle the past-key values correctly, but please let me know if there's still an issue :)

lewtun reviewed (commented):
Thank you for adding ONNX support for this popular model @pramodith! 

This looks like a great start and I think we can try copying over the prior work done with the BART export to see if that fixes up the issues with exporting with past key-value pairs

pramodith committed:
Changing to the pegasus-base-x model for testing in test_onnx_v2.py to since the model is smaller. Copying code from configuration_bart.py and adding to configuration_pegasus.py to resolve the past_key_values issue.

pramodith committed:
Merge remote-tracking branch 'origin/pramodith_pegasus_onnx_config' into pramodith_pegasus_onnx_config

# Conflicts:
#	src/transformers/models/pegasus/configuration_pegasus.py

pramodith committed:
Merge branch 'main' into pramodith_pegasus_onnx_config

pramodith committed:
Changing to the pegasus-base-x model for testing in test_onnx_v2.py to since the model is smaller. Copying code from configuration_bart.py and adding to configuration_pegasus.py to resolve the past_key_values issue.

pramodith committed:
Merge remote-tracking branch 'origin/pramodith_pegasus_onnx_config' into pramodith_pegasus_onnx_config

# Conflicts:
#	src/transformers/models/pegasus/configuration_pegasus.py

pramodith committed:
Undoing changes to serialization.mdx

pramodith committed:
Undoing changes to serialization.mdx

pramodith committed:
Fixing formatting issue.

pramodith committed:
Fixing formatting issue.

NielsRogge committed:
[VideoMAE] Improve code examples (#18919)

* Simplify code example

* Add seed

ydshieh committed:
Add checks for more workflow jobs (#18905)

* add check for scheduled CI

* Add check to other CIs

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

nbroad1881 committed:
Accelerator end training (#18910)

* add accelerator.end_training()

Some trackers need this to end their runs.

* fixup and quality

* add space

* add space again ?!?

sywangyi committed:
update the train_batch_size in case HPO change batch_size_per_device (#18918)

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

Rocketknight1 committed:
Update TF fine-tuning docs (#18654)

* Update TF fine-tuning docs

* Fix formatting

* Add some section headers so the right sidebar works better

* Squiggly it

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/training.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Explain things in the text, not the comments

* Make the two dataset creation methods into a list

* Move the advice about collation out of a <Tip>

* Edits for clarity

* Edits for clarity

* Edits for clarity

* Replace `to_tf_dataset` with `prepare_tf_dataset` in the fine-tuning pages

* Restructure the page a little bit

* Restructure the page a little bit

* Restructure the page a little bit

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

gante committed:
TF: final bias as a layer in seq2seq models (replicate TFMarian fix) (#18903)

ydshieh committed:
remvoe `_create_and_check_torch_fx_tracing` in specific test files (#18667)

* remvoe _create_and_check_torch_fx_tracing defined in specific model test files

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

tjruwase committed:
[DeepSpeed ZeRO3] Fix performance degradation in sharded models (#18911)

* [DeepSpeed] Fix performance degradation in sharded models

* style

* polish

Co-authored-by: Stas Bekman <stas@stason.org>

ankrgyl committed:
Add DocumentQuestionAnswering pipeline (#18414)

* [WIP] Skeleton of VisualQuestionAnweringPipeline extended to support LayoutLM-like models

* Fixup

* Use the full encoding

* Basic refactoring to DocumentQuestionAnsweringPipeline

* Cleanup

* Improve args, docs, and implement preprocessing

* Integrate OCR

* Refactor question_answering pipeline

* Use refactored QA code in the document qa pipeline

* Fix tests

* Some small cleanups

* Use a string type annotation for Image.Image

* Update encoding with image features

* Wire through the basic docs

* Handle invalid response

* Handle empty word_boxes properly

* Docstring fix

* Integrate Donut model

* Fixup

* Incorporate comments

* Address comments

* Initial incorporation of tests

* Address Comments

* Change assert to ValueError

* Comments

* Wrap `score` in float to make it JSON serializable

* Incorporate AutoModeLForDocumentQuestionAnswering changes

* Fixup

* Rename postprocess function

* Fix auto import

* Applying comments

* Improve docs

* Remove extra assets and add copyright

* Address comments

Co-authored-by: Ankur Goyal <ankur@impira.com>

ydshieh committed:
pin TF 2.9.1 for self-hosted CIs (#18925)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ymwangg committed:
Fix XLA fp16 and bf16 error checking (#18913)

* Fix XLA fp16 and bf16 error checking

* Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

colindean committed:
Starts on a list of external deps required for dev (#18929)

* Starts on a list of external deps required for dev

I've found that I need to install MeCab manually on my AS Mac.

* Generalizes OS nascent dependency list

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

lewtun committed:
Add image height and width to ONNX dynamic axes (#18915)

stevhliu committed:
Skip some doctests in quicktour (#18927)

* skip some code examples for doctests

* make style

* fix code snippet formatting

* separate code snippet into two blocks

Devlee247 committed:
Fix LayoutXLM wrong link in README (#18932)

* fix LayoutXLM wrong link in README

* fix LayoutXLM worng link in index.mdx

NielsRogge committed:
Add X-CLIP (#18852)

* First draft

* Improve conversion script

* Make vision encoder work

* More improvements

* Improve conversion script

* Fix quality

* Add MultiframeIntegrationTransformer

* More improvements

* Make MiT output work

* Fix quality

* Add prompts generator

* Add tests

* Fix some tests

* Fix some more tests

* Fix more tests

* Improve conversion script

* Fix model outputs

* Fix more tests

* Add XClipProcessor

* Use processor in conversion script

* Fix integration test

* Update README, fix docs

* Fix all tests

* Add MIT output to XClipOutput

* Create better variable names

* Rename XClip to XCLIP

* Extend conversion script

* Add support for large models

* Add support for 16 frame models

* Add another model'

* Fix module issue

* Apply suggestions from code review

* Add figure to docs

* Fix CLIPProcessor issue

* Apply suggestions from code review

* Delete file

* Convert more checkpoints

* Convert last checkpoint

* Update nielsr to microsoft

NimaBoscarino committed:
Update translation requests contact (#18941)

* Update TRANSLATING.md

Update the contact to @GuggerSylvain

* Update docs/TRANSLATING.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

nghuyong committed:
add task_type_id to BERT to support ERNIE-2.0 and ERNIE-3.0 models (#18686)

* add_ernie

* remove Tokenizer in ernie

* polish code

* format code style

* polish code

* fix style

* update doc

* make fix-copies

* change model name

* change model name

* fix dependency

* add more copied from

* rename ErnieLMHeadModel to ErnieForCausalLM
do not expose ErnieLayer
update doc

* fix

* make style

* polish code

* polish code

* fix

* fix

* fix

* fix

* fix

* final fix

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sanchit-gandhi committed:
[JAX] Replace all jax.tree_* calls with jax.tree_util.tree_* (#18361)

* [JAX] Replace all jax.tree_* calls with jax.tree_util.tree_*

* fix double tree_util

Raalsky committed:
Neptune.ai integration improvements (#18934)

* NeptuneCallback improvements

* After review suggestions and deduplication of initial run

* Added volatile checkpoints support due to missing post-rebase commit

* Update README per review comments

- Remove list formatting
- Correct Neptune docs link

Co-authored-by: Sabine <sabine.nyholm@neptune.ai>

ekagra-ranjan committed:
Generate: Simplify is_pad_token_not_equal_to_eos_token_id (#18933)

Rocketknight1 committed:
Fix train_step, test_step and tests for CLIP (#18684)

* Fix train_step and test_step, correctly enable CLIP fit test

* Stop using get_args on older Python versions

* Don't use get_origin either

* UnionType is actually even newer, don't use that either

* Apply the same fix to test_loss_computation

* Just realized I was accidentally skipping a bunch of tests!

* Fix test_loss_computation for models without separable labels

* Fix scalar losses in test_step and train_step

* Stop committing your breakpoints

* Fix Swin loss shape

* Fix Tapas loss shape

* Shape fixes for TAPAS, DeIT, HuBERT and ViTMAE

* Add loss computation to TFMobileBertForPreTraining

* make fixup and move copied from statement

* make fixup and move copied from statement

* Correct copied from

* Add labels and next_sentence_label inputs to TFMobileBERT

* Make sure total_loss is always defined

* Update tests/test_modeling_tf_common.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Fix copied from

* Ensure CTC models get labels in tests

* Ensure CTC models get labels in tests

* Fix tests for vit_mae

* Fix tests for vit_mae

* Fix tests for vit_mae

* Reduce batch size for wav2vec2 testing because it was causing OOM

* Skip some TAPAS tests that are failing

* Skip a failing HuBERT test

* make style

* Fix mobilebertforpretraining test

* Skip Wav2Vec2 tests that use huge amounts of mem

* Skip keras_fit for Wav2Vec2 as well

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

sgugger committed:
Exit early in load if no weights are in the sharded state dict (#18937)

Commit:
update black target version (#18955)

* update black target version

* add comment

as per https://github.com/huggingface/transformers/pull/18955#issuecomment-1242081649

* revert change

Will only update to 3.7 after black 2023 upgrade in January

gante committed:
RFC: Replace custom TF embeddings by Keras embeddings (#18939)

gante committed:
TF: unpin maximum TF version (#18917)

sgugger committed:
Revert "TF: unpin maximum TF version (#18917)" (#18972)

This reverts commit d8cf3b20875baee97f4bea64ffd17670aa57c37b.

pramodith committed:
Fixing formatting issue.
